activerecord/lib/active_record/load_tree.rb

```diff
@@ -0,0 +1,161 @@
+  # With this feature you can debug how a record was loaded, where it happened in the call stack
+  # and in the tree hierarchy and what else was loaded at the same time.
+  #
+  # See all Records loaded since the debug mode was turned on:
+  #
+  # AciveRecord::LoadTree.loaded_records
+  #
+  #
+  # It is then possible to iterate through these and inspect the full tree of each record,
+  # allowing you to see what was loaded and how it was loaded and debug full trees of records.
+  #
+  class LoadTree
+    attr_reader :association_name, :record_class_name, :loaded_associations,
+                :load_call_stack, :siblings, :parent
+    attr_writer :loaded_association
+    require "weakref"
+
+    # Mark an ActiveRecord instance/s as loaded.
+    def self.add_loaded_records(records)
+      return loaded_records unless ActiveRecord.dynamic_includes_overfetch_debugging?
+      return loaded_records if records.nil?
+      records = Array(records) unless records.is_a?(Array)
+      loaded_records.concat(records).uniq!
+    end
+
+    # All ActiveRecords instances marked as loaded.
+    def self.loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] ||= []
+    end
+
+    # Reset the state of tracked loaded and used records.
+    def self.clear_loaded_records
+      ActiveSupport::IsolatedExecutionState[:active_record_loaded_records] = []
+    end
+
+    # Create a Load Tree instance representing the loaded record and its place in the overall
+    # hierarchy of the object tree.
+    #
+    # parent: ActiveRecord instance that is the parent of the loaded record.
+    # siblings: ActiveRecord instances that were loaded at the same time as the loaded record.
+    # association_name: The name of the association that was loaded.
+    def initialize(creator: nil, parent: nil, siblings: [], association_name: nil)
+      raise ArgumentError, "siblings must be an array" unless siblings.is_a?(Array)
+      raise ArgumentError, "Cannot be in an association if it is a root element" if parent.nil? && !association_name.nil?
+      raise ArgumentError, "all siblings must be have a load tree" unless siblings.all? { |r| r.respond_to?(:_load_tree) && r._load_tree.is_a?(LoadTree) }
+      set_parent(parent)
+
+      @record_class_name = creator.class.name
+      @association_name = association_name
+      parent._load_tree.add_loaded_association(association_name) unless parent.nil? || association_name.nil?
+      @loaded_associations = []
+      @load_call_stack = caller if ActiveRecord.dynamic_includes_overfetch_debugging?
+      reset_siblings(siblings)
+    end
+
+    def reset_siblings(siblings)
+      # We only want to keep sibling of the same class type as the creator
+      # This ensures we will not try to preload any associations that may not
+      # exist on some of the sibling which were loaded because of STI.
+      @siblings = siblings.select { |s| s.class.name == @record_class_name }.compact
```

Comment:
I was storing these as WeakRefs, but after a conversation with @jhawthorn I switched back to just directly referencing them.

Reply:
I still think these should be weakrefs. Querying on behalf of an object that's no longer accessible from the live object graph seems pretty clearly wasteful?